src/api/analytics/gtag.js

```diff
@@ -0,0 +1,192 @@
+
+const oneOffExperimentDownload = (accessionCode) =>
+  event('one_off_experiment_download', {
+    one_off_experiment_download: accessionCode
+  })
+
+const regeneratedDataset = (state, defaultOptions, newOptions) => {
+  // due to GA char limit, keys names are abbreviated
+  const format = (option) => {
+    return `A: ${option.aggregate_by}, T: ${
+      transformation[option.scale_by]
+    }, QN: ${option.quantile_normalize ? 'Not skipped' : 'Skipped'}`
+  }
+
+  event('regenerated_dataset', {
+    regenerated_state: state ? 'Expired' : 'Valid',
+    regenerated_download_options: `${format(defaultOptions)} ${
+      newOptions ? `(new) ${format(newOptions)}` : ''
+    }`
+  })
+}
+
+const sharedDataset = (status) =>
+  event('shared_dataset', {
+    shared_state: status ? 'Processed' : 'Unprocessed'
+  })
+
+/* --- Links --- */
+const experimentPageClick = (from) =>
+  event('page_view', {
+    experiment_page_click_from: from
+  })
+
+const exploredUsageClick = (usage, type = 'dataset') =>
+  event(`click`, { [`${type}_explored_usage`]: usage })
+
+const navClick = (item, type = 'global') =>
+  event(`page_view`, {
+    [`${type}_nav_item`]: item
+  })
+
+const outboundClick = (url) => {
+  let parameter
+  switch (true) {
+    case url === links.alsf || url === links.ccdl:
+      parameter = 'alsf_domain_name'
+      break
+    case url === links.ccdl_donate:
+      parameter = 'donate_click_from'
+      break
+    case url.includes(links.refinebio_githubio):
+      parameter = 'alsf_githubio_click_from'
+      break
+    case url.includes(links.alsf_github):
+      parameter = 'alsf_repo_click_from'
+      break
+    case url.includes(links.refinebio_api_docs):
+      parameter = 'api_docs_click_from'
+      break
+    case url.includes(links.refinebio_docs):
+      parameter = 'docs_click_from'
+      break
+    case url.includes(links.ccdl_twitter):
+      parameter = 'social_media'
+      break
+    default:
+      break
+  }
+
+  event('click', {
+    [parameter]: url
+  })
+}
+
+/* --- Search --- */
+const filterCombination = (facets, query) => {
+  const filters = getSortedFilters(facets, query)
+  const combination = getFilterCombination(filters)
+
+  event('page_view', { filter_combination: combination })
+}
+
+const filterType = (type) => event('filter_type', { filter_type: type })
+
+const toggleFilterItem = (checked, item) =>
+  event('toggled_filter_item', {
+    toggled_filter_item: `${checked ? 'Add' : 'Remove'} - ${item}`
+  })
+
+const searchTerm = (term) => event('search_text', { search_text: term })
+
+// helpers
+const getFilterCombination = (sortedFilters) => {
+  const { keys, filters } = sortedFilters
+  let formattedFilters = ''
+
+  // due to GA char limit, keys names are abbreviated
+  for (const x of keys) {
+    if (filters[x]) {
+      formattedFilters = `${formattedFilters}[${
+        x === keys[0] ? 'O' : x[0].toUpperCase()
+      }]:${filters[x]} `
+    }
+  }
+
+  return formattedFilters
```

Comment:
```suggestion
  // due to GA char limit, keys names are abbreviated
  return keys
   .filter((key) => filters[key])
   .map((key) => `[${key[0].toUpperCase()}]:${filters[key]}`)   
   .join(" ")
```

I'm a bit perplexed by this condition, I see from the comment that we are trying to truncate it. But the issue I see is if the first `key` is not in the filter, then we would never return `O` and always return `x[0].toUpperCase()` ?
```javascript
x === keys[0] ? 'O' : x[0].toUpperCase()
```

Reply:
Oh yes, this is because the property name for the downloadable organisms is returned as `downloadable_organism_names` from the BE API (https://api.refine.bio/v1/#operation/search_list) which is then shortened to `downloadable_organism` on the client.

In the UI, we display those downloadable organisms in the "Organism" section of the facets:
![Screenshot 2024-08-15 at 6 52 05 PM](https://github.com/user-attachments/assets/37399291-c46c-4605-82f3-89fd8eb2c806)

For this reason, if items from the "Organism" section are selected, it will be explicitly set to `O` (otherwise `D`) to match the UI.

e.g., )  The selected filters may be as follows:

```js
sortedFilters = {
      // the first item will be explicitly set to 'O' 
      keys: ['downloadable_organism', 'platform', 'technology'], 
      filters: {
        downloadable_organism: ['Homo sapiens'],
        platform: 'Illumina HiSeq 2500',
        technology: ['Microarray', 'RNA-seq']
      }
    }
 ```

In analytics, the above filter combination will be collected and displayed in the "Filter Combination" report as follows:

`[O]:Homo sapiens [P]:Illumina HiSeq 2500 [T]:Microarray,RNA-seq`

Filter Combination Report in `CCDL > Refine.bio Web -Dev` (to see the demo data, please set the custom date to `May 1, 2023` to present)

![Screenshot 2024-08-15 at 7 43 22 PM](https://github.com/user-attachments/assets/2b3662e1-d6c6-4efd-888d-2c1aabb43fb7)



This data structure was discussed previously and added to the google sheet [Custom Events List > Search > Most used filter combination](https://docs.google.com/spreadsheets/d/1qKeq8_pEezvq-OAp6ohCjJhrr9EQYLLyGYFAQe9XvS4/edit?gid=1996653253#gid=1996653253) some times ago (back in January..!). When you have a moment, please take another look and let me know your thoughts on this. Thank you, David!